gflownet/envs/amp.py

```diff
@@ -248,102 +261,68 @@ def state2policy(self, state=None):
-        #     state_policy[rows, np.concatenate(cols)] = 1.0
+        state_tensor = torch.vstack(states)
+        state_policy = self.statetorch2policy(state_tensor)
         return state_policy
 
     def statetorch2policy(
-        self, states: TensorType["batch", "state_dim"]
-    ) -> TensorType["batch", "policy_output_dim"]:
-        # TODO: prettify and verify logic
-        cols, lengths = zip(
-            *[
-                (
-                    (
-                        state[
-                            : torch.where(state == self.invalid_state_element)[0][0]
-                            if state[-1] == self.invalid_state_element
-                            else len(state)
-                        ].to(self.device)
-                        + torch.arange(
-                            len(
-                                state[
-                                    : torch.where(state == self.invalid_state_element)[
-                                        0
-                                    ][0]
-                                    if state[-1] == self.invalid_state_element
-                                    else len(state)
-                                ]
-                            )
-                        ).to(self.device)
-                        * self.n_alphabet
-                    ),
-                    torch.where(state == self.invalid_state_element)[0][0]
-                    if state[-1] == self.invalid_state_element
-                    else len(state),
-                )
-                for state in states
-            ]
-        )
-        lengths = torch.Tensor(list(lengths)).to(self.device).to(torch.int64)
-        cols = torch.cat(cols, dim=0).to(torch.int64).to(self.device)
-        rows = torch.repeat_interleave(
-            torch.arange(len(states)).to(self.device), lengths
+        self, states: TensorType["batch", "max_seq_length"]
+    ) -> TensorType["batch", "policy_input_dim"]:
+        if states.dtype != torch.long:
+            states = states.to(torch.long)
+        state_onehot = (
+            F.one_hot(states, self.n_alphabet + 2)[:, :, :-2]
+            .to(self.float)
+            .to(self.device)
         )
+        state_padding_mask = (states != self.padding_idx).to(self.float).to(self.device)
+        state_onehot_pad = state_onehot * state_padding_mask.unsqueeze(-1)
+        # Assertion works as long as [PAD] is last key in lookup table.
+        assert torch.eq(state_onehot_pad, state_onehot).all()
         state_policy = torch.zeros(
-            (len(states), self.n_alphabet * self.max_seq_length),
-            dtype=self.float,
-            device=self.device,
+            states.shape[0], self.max_seq_length, self.n_alphabet
         )
-        state_policy[rows, cols] = 1.0
-        return state_policy
+        state_policy[:, : state_onehot.shape[1], :] = state_onehot
+        return state_policy.reshape(states.shape[0], -1).to(self.device).to(self.float)
```

Comment:
Minor, but it is better to set `device` and `dtype` directly when creating the tensor (in `zeros` here) to avoid making unnecessary copies from cpu to gpu.